src/github.com/matrix-org/dendrite/clientapi/routing/login.go

```diff
@@ -57,6 +57,21 @@ func passwordLogin() loginFlows {
 	return f
 }
 
+// GetLocalpartDomainFromUserid extracts localpart & domain of server from userID
+// Returns JSON error in case of invalid username.
+func GetLocalpartDomainFromUserid(userID string) (
```

Comment:
firstly, it's not obvious to me that a function like this belongs in `routing.login`: the fact that we're calling it from somewhere that isn't about handling a /login request shows that.

But generally I don't think this is a nice interface. It's confusing that it returns an error-like thing that turns out not to actually be an `error`, and you haven't really saved that much code at the call site.

If you wanted to you *could* factor out something like:

```
func BuildInvalidUsernameResponse() {
    return util.JSONResponse{
        Code: 400,
        JSON: jsonerror.InvalidUsername("Invalid username"),
    }
}
```

... but it's really not obvious to me that is worthwhile.

In short I'd just keep this inline.